src/Entity/SnapshotManager.php

```diff
@@ -231,60 +233,40 @@ public function cleanup(PageInterface $page, $keep)
             throw new \RuntimeException(sprintf('Please provide an integer value, %s given', \gettype($keep)));
         }
 
-        $tableName = $this->getTableName();
-        $platform = $this->getConnection()->getDatabasePlatform()->getName();
-
-        if ('mysql' === $platform) {
-            return $this->getConnection()->exec(sprintf(
-                'DELETE FROM %s
-                WHERE
-                    page_id = %d
-                    AND id NOT IN (
-                        SELECT id
-                        FROM (
-                            SELECT id, publication_date_end
-                            FROM %s
-                            WHERE
-                                page_id = %d
-                            ORDER BY
-                                publication_date_end IS NULL DESC,
-                                publication_date_end DESC
-                            LIMIT %d
-                        ) AS table_alias
-                )',
-                $tableName,
-                $page->getId(),
-                $tableName,
-                $page->getId(),
-                $keep
-            ));
+        $innerQb = $this->getRepository()->createQueryBuilder('i');
+        $expr = $innerQb->expr();
+
+        // try a better Function expression for this?
+        $ifNullExpr = sprintf(
+            'CASE WHEN %s THEN 1 ELSE 0 END',
+            $expr->isNull('i.publicationDateEnd')
+        );
+
+        // Subquery DQL doesn't support Limit
+        $innerQb
+            ->select('DISTINCT i.id')
```

Comment:
Distinct is a MySQL function, I guess you can have the same result using group by